### PR TITLE
LPS-205115 Update notification link to redirect to System Settings and guarantee that notification to admin user is only sent once

### PR DESCRIPTION
**Pull request by thallescezar**
Related issue: https://liferay.atlassian.net/browse/LPD-11999

### CODE REVIEW

**Comment by thallescezar** [comment]
ci:test:sf

**Comment by thallescezar** [comment]
ci:test:object

**Comment by thallescezar** [comment]
ci:test:relevant

File: modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectEntryServiceImpl.java [hunk 1]
```diff
@@ -652,11 +655,26 @@ private void _sendUserNotificationEvents(ObjectDefinition objectDefinition)
 			objectDefinition.getCompanyId(), RoleConstants.ADMINISTRATOR);
 
 		for (long userId : _userLocalService.getRoleUserIds(role.getRoleId())) {
-			int count =
-				_userNotificationEventLocalService.
-					getUserNotificationEventsCount(
-						userId, portletId, timestamp, true,
-						"%\"exceedsObjectEntryLimit\":%");
+			List<UserNotificationEvent> userNotificationEventList =
```
Comment (guilhermedcamacho) [inline modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectEntryServiceImpl.java:658]: Please rename this variable to `userNotificationEvents`

File: modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectEntryServiceImpl.java [hunk 2]
```diff
@@ -652,11 +655,26 @@ private void _sendUserNotificationEvents(ObjectDefinition objectDefinition)
...
+			int count = 0;
+
+			for (UserNotificationEvent userNotificationEvent :
+					userNotificationEventList) {
+
+				if (count > 0) {
+					continue;
+				}
+
+				JSONObject payloadJSONObject = _jsonFactory.createJSONObject(
```
Comment (guilhermedcamacho) [inline modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectEntryServiceImpl.java:671]: Please rename this variable to `jsonObject`

File: modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectEntryServiceImpl.java [hunk 3]
```diff
@@ -652,11 +655,26 @@ private void _sendUserNotificationEvents(ObjectDefinition objectDefinition)
...
+					userNotificationEvent.getPayload());
+
+				if (payloadJSONObject.get("exceedsObjectEntryLimit") != null) {
+					count++;
```
Comment (guilhermedcamacho) [inline modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectEntryServiceImpl.java:675]: I think it's better to use `jsonObject.has("exceedsObjectEntryLimit")` here.

File: modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectEntryServiceImpl.java [hunk 4]
```diff
@@ -652,11 +655,26 @@ private void _sendUserNotificationEvents(ObjectDefinition objectDefinition)
...
+			List<UserNotificationEvent> userNotificationEventList =
+				_userNotificationEventLocalService.getUserNotificationEvents(
+					userId, portletId, timestamp, true);
+
+			int count = 0;
+
+			for (UserNotificationEvent userNotificationEvent :
+					userNotificationEventList) {
+
+				if (count > 0) {
```
Comment (guilhermedcamacho) [inline modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectEntryServiceImpl.java:667]: What do you think to change this whole logic to the code below?

```
		for (long userId : _userLocalService.getRoleUserIds(role.getRoleId())) {
			List<UserNotificationEvent> userNotificationEvents =
				_userNotificationEventLocalService.getUserNotificationEvents(
					userId, portletId, timestamp, true);

			boolean notificationReceived = false;

			for (UserNotificationEvent userNotificationEvent :
					userNotificationEvents) {

				JSONObject jsonObject = _jsonFactory.createJSONObject(
					userNotificationEvent.getPayload());

				if (jsonObject.has("exceedsObjectEntryLimit")) {
					notificationReceived = true;
					break;
				}
			}

			if (!notificationReceived) {
				userIds.add(userId);
			}
		}
```

File: modules/apps/object/object-test/src/testIntegration/java/com/liferay/object/service/test/ObjectEntryServiceTest.java [hunk 5]
```diff
@@ -849,15 +853,32 @@ private void _assertUserNotificationEventsCount() throws PortalException {
 		long[] userIds = _userLocalService.getRoleUserIds(role.getRoleId());
 
 		for (long userId : userIds) {
-			int count =
-				_userNotificationLocalService.getUserNotificationEventsCount(
+			List<UserNotificationEvent> userNotificationEventList =
```
Comment (guilhermedcamacho) [inline modules/apps/object/object-test/src/testIntegration/java/com/liferay/object/service/test/ObjectEntryServiceTest.java:856]: Please apply the same changes requested above to the test class.

**Review by guilhermedcamacho — CHANGES_REQUESTED** [review]

File: modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectEntryServiceImpl.java (hunk 4 above)
Comment (thallescezar) [reply modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectEntryServiceImpl.java:667]: I thought that using "break" could be a problem. So I used other way of doing this loop. If it is not a problem I can use it.

Comment (guilhermedcamacho) [reply modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectEntryServiceImpl.java:667]: It's not a problem, you can use.

**Comment by thallescezar** [comment]
ci:test:sf

**Comment by thallescezar** [comment]
ci:test:object

**Comment by thallescezar** [comment]
ci:test:relevant

**Comment by thallescezar** [comment]
ci:report:2531541969

**Comment by thallescezar** [comment]
ci:report:2531579050

**Comment by thallescezar** [comment]
Errors on tests are not related.

File: modules/apps/object/object-test/src/testIntegration/java/com/liferay/object/service/test/ObjectEntryServiceTest.java [hunk 6]
```diff
@@ -849,17 +853,34 @@ private void _assertUserNotificationEventsCount() throws PortalException {
...
-					true, "%\"exceedsObjectEntryLimit\":%");
+					true);
 
-			Assert.assertTrue(count == 1);
+			int count = 0;
+
+			for (UserNotificationEvent userNotificationEvent :
+					userNotificationEvents) {
+
+				if (count > 0) {
```
Comment (guilhermedcamacho) [inline modules/apps/object/object-test/src/testIntegration/java/com/liferay/object/service/test/ObjectEntryServiceTest.java:871]: Hey @thallescezar, correct me if I'm wrong, but if you keep this condition with the continue statement, the count variable will never reach any number greater than 1, so you're not testing the case/possible bug when the admin user gets 2 or more notifications.

**Review by guilhermedcamacho — CHANGES_REQUESTED** [review]

**Comment by thallescezar** [comment]
ci:test:sf

**Comment by guilhermedcamacho** [comment]
ci:forward:force

**Comment by guilhermedcamacho** [comment]
I'm appending one more commit to fix the semantic versioning.
Manually forwarded to [brianchandotcom#146735](https://github.com/brianchandotcom/liferay-portal/pull/146735)

**Pull request closed without merging**